harness/determined/_info.py

```diff
@@ -0,0 +1,318 @@
+        # in their training code.  They should focus on the data field and the hparams field, which
+        # have always been and will always be user-defined.
+        self._config = config
+
+        # rb: These fields are private because I am pretty confident that we need to find better
+        # ways to pass them around the system.  But for now, they're passed in as environment
+        # variables and for now we have to be able to handle that.
+        # TODO: decide if we really want to track latest_batch for users or not.
+        self._latest_batch = latest_batch
+        # TODO: get rid of trial_run_id in favor of allocation id.
+        self._trial_run_id = trial_run_id
+        # TODO: decide if the experiment config is the right place for users to set a debug flag.
+        self._debug = debug
+        # TODO: is this derivable from the slot ids?
+        self._unique_port_offset = unique_port_offset
+        # TODO: Get rid of this in favor of launch layer configs?
+        self._inter_node_network_interface = inter_node_network_interface
+
+    @classmethod
+    def _from_env(cls) -> "TrialInfo":
+        assert "DET_TRIAL_ID" in os.environ, "must be run inside a Trial container"
+        experiment_config = json.loads(os.environ["DET_EXPERIMENT_CONFIG"])
+        return cls(
+            trial_id=int(os.environ["DET_TRIAL_ID"]),
+            experiment_id=int(os.environ["DET_EXPERIMENT_ID"]),
+            trial_seed=int(os.environ["DET_TRIAL_SEED"]),
+            hparams=json.loads(os.environ["DET_HPARAMS"]),
+            config=experiment_config,
+            latest_batch=int(os.environ["DET_LATEST_BATCH"]),
+            trial_run_id=int(os.environ["DET_TRIAL_RUN_ID"]),
+            debug=experiment_config.get("debug", False),
+            unique_port_offset=int(os.environ["DET_UNIQUE_PORT_OFFSET"]),
+            inter_node_network_interface=os.environ.get("DET_INTER_NODE_NETWORK_INTERFACE"),
+        )
+
+    def _to_file(self, path: str = "/run/determined/info/trial.json") -> None:
+        with open(path, "w") as f:
+            json.dump({k.lstrip("_"): v for k, v in vars(self).items()}, f)
+
+    @classmethod
+    def _from_file(cls, path: str = "/run/determined/info/trial.json") -> Optional["TrialInfo"]:
+        if not os.path.exists(path):
+            return None
+        with open(path) as f:
+            return cls(**json.load(f))
+
+
+class ResourcesInfo:
+    def __init__(self, gpu_uuids: List[str]) -> None:
+        self._gpu_uuids = gpu_uuids
+
+    @property
+    def gpu_uuids(self) -> List[str]:
+        return self._gpu_uuids
+
+    @classmethod
+    def _by_inspection(cls) -> "ResourcesInfo":
+        return cls(gpu_uuids=gpu.get_gpu_uuids())
+
+    def _to_file(self, path: str = "/run/determined/info/resources.json") -> None:
+        with open(path, "w") as f:
+            json.dump({k.lstrip("_"): v for k, v in vars(self).items()}, f)
```

Comment:
would it be helpful to have a to_json classmethod to serialize?

Reply:
I can add a helper function, but it will be a standalone function.  I think that inheritance hierarchies which only exist for code sharing purposes result in messy object-oriented code:  E.g. a TrialInfo is not "like" a ClusterInfo object.